{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.\n",
      "  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')\n",
      "/opt/conda/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.\n",
      "  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio import AlignIO\n",
    "from Bio.Phylo.Applications import FastTreeCommandline\n",
    "from Bio import Phylo\n",
    "%matplotlib inline\n",
    "import scipy.cluster.hierarchy as cl\n",
    "import scipy.spatial.distance as ssd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import Entrez, SeqIO\n",
    "import time\n",
    "from termcolor import colored\n",
    "\n",
    "Entrez.email = \"vincent.deruaz@master.hes-so.ch\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class sea_inphinity():\n",
    "    def __init__(self, db_name):\n",
    "        self.db_name = db_name\n",
    "        self.connection()\n",
    "        self.verbose = True\n",
    "        self.out_dir = './out/'\n",
    "        \n",
    "    def print_(self, string):\n",
    "        if self.verbose:\n",
    "            print(string)\n",
    "        \n",
    "    def connection(self):\n",
    "        # Connect to the database\n",
    "        self.connection = pymysql.connect(host='localhost',\n",
    "                                     user='root',\n",
    "                                     password='',\n",
    "                                     db=self.db_name,\n",
    "                                     charset='utf8mb4',\n",
    "                                     cursorclass=pymysql.cursors.DictCursor)\n",
    "        \n",
    "    def simple_execute(self, sql, verbose):\n",
    "        try:\n",
    "            with self.connection.cursor() as cursor:\n",
    "                res = cursor.execute(sql)\n",
    "\n",
    "                if verbose:\n",
    "                    self.print_(res)\n",
    "                    \n",
    "                return res\n",
    "\n",
    "            self.connection.commit()\n",
    "        finally:\n",
    "            pass\n",
    "        \n",
    "    def get_list_pham(self, limit):\n",
    "        cur = self.connection.cursor()\n",
    "\n",
    "        if limit == -1:\n",
    "            sql = \"SELECT * FROM pham\"\n",
    "        else:\n",
    "            sql = \"SELECT * FROM pham LIMIT %d\" % (limit)\n",
    "        cur.execute(sql)\n",
    "\n",
    "        for row in cur:\n",
    "            self.print_(row)\n",
    "\n",
    "        return cur\n",
    "        \n",
    "    \n",
    "    def get_list_name_pham(self, limit):\n",
    "        cur = self.connection.cursor()\n",
    "\n",
    "        if limit == -1:\n",
    "            sql = \"SELECT name FROM pham GROUP BY name\"\n",
    "        else:\n",
    "            sql = \"SELECT name FROM pham GROUP BY name LIMIT %d\" % (limit)\n",
    "        cur.execute(sql)\n",
    "        \n",
    "        list_name = []\n",
    "        for name in list(cur):\n",
    "            list_name.append(name['name'])\n",
    "\n",
    "        return list_name\n",
    "    \n",
    "    def get_specific_pham(self, name):\n",
    "        cur = self.connection.cursor()\n",
    "\n",
    "        cur.execute(\"SELECT GeneID FROM pham WHERE name = %s\", (name))\n",
    "\n",
    "        return cur\n",
    "    \n",
    "    def get_list_genes(self, gene_ids, verbose):\n",
    "        cur = self.connection.cursor()\n",
    "        \n",
    "        sql = \"SELECT * FROM gene WHERE GeneID IN ('%s');\" % ( \"','\".join(gene_ids))\n",
    "        \n",
    "        cur.execute(sql)\n",
    "\n",
    "        if verbose:\n",
    "            for row in cur:\n",
    "                self.print_(row)\n",
    "\n",
    "        return list(cur.fetchall())\n",
    "    \n",
    "    def get_phage(self, phage_id, verbose):\n",
    "        cur = self.connection.cursor()\n",
    "        \n",
    "        sql = \"SELECT * FROM phage WHERE PhageID = '%s';\" % (phage_id)\n",
    "        \n",
    "        cur.execute(sql)\n",
    "\n",
    "        if verbose:\n",
    "            for row in cur:\n",
    "                self.print_(row)\n",
    "\n",
    "        return list(cur.fetchall())\n",
    "    \n",
    "    def get_phage_from_gene_id(self, gene_id):\n",
    "        phage = sea_inphinity.get_phage(sea_inphinity.get_list_genes([gene_id], False)[0]['PhageID'], False)[0]\n",
    "        return phage\n",
    "    \n",
    "    def build_tree(self,pham):\n",
    "        genes = self.get_genes_from_a_pham(pham)\n",
    "        self.create_fasta(genes)\n",
    "        self.align_muscle()\n",
    "        self.compute_tree()\n",
    "        self.prepare_tree_fig()\n",
    "    \n",
    "    def get_genes_from_a_pham(self, pham):\n",
    "        pham_list = []\n",
    "        pham = sea_inphinity.get_specific_pham(pham)\n",
    "        for id in pham:\n",
    "            pham_list.append(id['GeneID'])\n",
    "            self.print_(id)\n",
    "        #print(pham_list)\n",
    "        genes = sea_inphinity.get_list_genes(pham_list, False)\n",
    "        \n",
    "        return genes\n",
    "        \n",
    "    def create_fasta(self, genes):\n",
    "        print('Creation of the FASTA file')\n",
    "        fasta = open(\"%sfasta.fa\" % (self.out_dir), \"w\")\n",
    "        self.print_(\"Number of Genes: %d\" % (len(genes)))\n",
    "        for gene in genes:\n",
    "            GeneID = gene['GeneID']\n",
    "            name = gene['Name']\n",
    "            description = \">%s - %s\" % (GeneID, name)\n",
    "\n",
    "            translation = gene['translation']\n",
    "\n",
    "            self.print_(description)\n",
    "            self.print_(translation)\n",
    "\n",
    "            fasta.write(description)\n",
    "            fasta.write('\\n')\n",
    "            fasta.write(translation)\n",
    "            fasta.write('\\n')\n",
    "\n",
    "        fasta.close()\n",
    "        \n",
    "    def align_muscle(self):\n",
    "        print('Alignment with MUSCLE')\n",
    "        muscle_loc = r'/home/pa/work/muscle3.8.31_i86linux64' # modifier si n√©cessaire\n",
    "\n",
    "        muscle_cline = MuscleCommandline(cmd=muscle_loc,input='%sfasta.fa' % (self.out_dir),out='%sout.aln' % (self.out_dir),clwstrict=True)\n",
    "        stdout, stderr = muscle_cline()\n",
    "\n",
    "        muscle_align = AlignIO.read('%sout.aln' % (self.out_dir),'clustal')\n",
    "        self.print_(muscle_align)\n",
    "        \n",
    "    def compute_tree(self):\n",
    "        print('Compute tree')\n",
    "        AlignIO.convert('%sout.aln' % (self.out_dir),'clustal','%sintermediate.phy' % (self.out_dir), 'phylip-relaxed')\n",
    "\n",
    "        cmd_fasttree = r'fasttree'\n",
    "        fasttree_cmdline = FastTreeCommandline(cmd=cmd_fasttree,fastest=True, \\\n",
    "                                               input='%sintermediate.phy' % (self.out_dir),out='%stree.tre' % (self.out_dir))\n",
    "        out_log, err_log = fasttree_cmdline()\n",
    "\n",
    "        self.print_('Out Log:')\n",
    "        self.print_(out_log)\n",
    "\n",
    "        self.print_('Error Log')\n",
    "        self.print_(err_log)\n",
    "        \n",
    "        self.tree = Phylo.read('%stree.tre' % (self.out_dir), 'newick')\n",
    "        \n",
    "    def prepare_tree_fig(self):\n",
    "        print('Prepare Tree Fig')\n",
    "        self.dmat = []\n",
    "        self.leaves = [str(cladit) for k,cladit in enumerate(self.tree.get_terminals())]\n",
    "        for l1,leave1 in enumerate(self.leaves):\n",
    "            d = []\n",
    "            for l2,leave2 in enumerate(self.leaves):\n",
    "                d.append(self.tree.distance(leave1,leave2))\n",
    "            self.dmat.append(d)\n",
    "            \n",
    "        \n",
    "        \n",
    "    def draw_tree(self):\n",
    "        print('Draw Tree')\n",
    "        Z = cl.linkage(ssd.squareform(self.dmat),method='average',metric='euclidean')\n",
    "        fig = plt.figure(num=None,figsize=(30,25),dpi=250)\n",
    "        dendro=cl.dendrogram(Z,labels=self.leaves,color_threshold=0.06,leaf_rotation=90,leaf_font_size=10)\n",
    "        \n",
    "        self.plt_tree = plt\n",
    "        \n",
    "        self.plt_tree.show()\n",
    "        \n",
    "    def print_informations_on_phage(self, gene_id):\n",
    "        phage = sea_inphinity.get_phage_from_gene_id(gene_id)\n",
    "        color = 'blue'\n",
    "        \n",
    "        source = phage['HostStrain']\n",
    "        if source == 'None':\n",
    "            color = 'red'\n",
    "            \n",
    "            if phage['Accession'] != '':\n",
    "                phage_ = phage['Accession']\n",
    "            else:\n",
    "                color = 'yellow'\n",
    "                phage_ = phage['Name'].split('-')[0]\n",
    "            \n",
    "            source = sea_inphinity.get_host_from_genbank(phage_)\n",
    "            print(source)\n",
    "        \n",
    "        print(\"******************************** \\\n",
    "            \\nGene ID: %s \\nPhage ID: %s \\nPhage Name: %s \\nPhage Accession: %s\" \\\n",
    "            % (gene_id, phage['PhageID'], phage['Name'], phage['Accession']))\n",
    "        print(colored(\"HostStrin: %s\" % (source), color))        \n",
    "        print(\"********************************\")\n",
    "                  \n",
    "                    \n",
    "        \n",
    "    def print_informations_on_phages(self, gene_ids):\n",
    "        print('===========================')\n",
    "        print(colored('Blue: Host come from the original database', 'blue'))\n",
    "        print(colored('Red: Host come from GenBank', 'red'))\n",
    "        print(colored('Yellow: DRAFT', 'yellow'))\n",
    "        print('===========================')\n",
    "        \n",
    "        for gene_id in gene_ids:\n",
    "            self.print_informations_on_phage(gene_id)\n",
    "            \n",
    "    def get_host_from_genbank(self, genome_id):\n",
    "        \n",
    "        try:\n",
    "            record = Entrez.efetch(db=\"nuccore\", id=genome_id, rettype=\"gb\", retmode=\"text\")\n",
    "\n",
    "            filename = 'out/genBankRecord.gb'\n",
    "            with open(filename, 'w') as f:\n",
    "                f.write(record.read())\n",
    "            parsed_gb_file = next(SeqIO.parse(filename, \"genbank\"))\n",
    "\n",
    "            #print(parsed_gb_file)\n",
    "            return parsed_gb_file.annotations[\"source\"]\n",
    "        \n",
    "        except:\n",
    "            return 'Not Found'\n",
    "            \n",
    "\n",
    "    \n",
    "sea_inphinity = sea_inphinity('sea')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Listing Pham"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "list_name = sea_inphinity.get_list_name_pham(-1)\n",
    "print(list_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing on Pham 2799"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creation of the FASTA file\n",
      "Alignment with MUSCLE\n"
     ]
    },
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 127 from '/home/pa/work/muscle3.8.31_i86linux64 -in ./out/fasta.fa -out ./out/out.aln -clwstrict', message '/bin/sh: 1: /home/pa/work/muscle3.8.31_i86linux64: not found'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m\u001b[0m",
      "\u001b[1;31mApplicationError\u001b[0mTraceback (most recent call last)",
      "\u001b[1;32m<ipython-input-4-eb02f741a639>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0msea_inphinity\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mverbose\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0msea_inphinity\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbuild_tree\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'2799'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0msea_inphinity\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdraw_tree\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-2-ca3341925f1f>\u001b[0m in \u001b[0;36mbuild_tree\u001b[1;34m(self, pham)\u001b[0m\n\u001b[0;32m    103\u001b[0m         \u001b[0mgenes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_genes_from_a_pham\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpham\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    104\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcreate_fasta\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgenes\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 105\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0malign_muscle\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    106\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcompute_tree\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    107\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_tree_fig\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-2-ca3341925f1f>\u001b[0m in \u001b[0;36malign_muscle\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    144\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    145\u001b[0m         \u001b[0mmuscle_cline\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mMuscleCommandline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmuscle_loc\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'%sfasta.fa'\u001b[0m \u001b[1;33m%\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mout_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mout\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'%sout.aln'\u001b[0m \u001b[1;33m%\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mout_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mclwstrict\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 146\u001b[1;33m         \u001b[0mstdout\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmuscle_cline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    147\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    148\u001b[0m         \u001b[0mmuscle_align\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'%sout.aln'\u001b[0m \u001b[1;33m%\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mout_dir\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'clustal'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/opt/conda/lib/python3.5/site-packages/Bio/Application/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[0;32m    515\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    516\u001b[0m             raise ApplicationError(return_code, str(self),\n\u001b[1;32m--> 517\u001b[1;33m                                    stdout_str, stderr_str)\n\u001b[0m\u001b[0;32m    518\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    519\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mApplicationError\u001b[0m: Non-zero return code 127 from '/home/pa/work/muscle3.8.31_i86linux64 -in ./out/fasta.fa -out ./out/out.aln -clwstrict', message '/bin/sh: 1: /home/pa/work/muscle3.8.31_i86linux64: not found'"
     ]
    }
   ],
   "source": [
    "sea_inphinity.verbose = False\n",
    "sea_inphinity.build_tree('2799')\n",
    "sea_inphinity.draw_tree()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(sea_inphinity.tree.get_terminals())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Clade(branch_length=0.02764, name='663557_12')\n",
    "Clade(branch_length=0.0, name='Backyardigan-DRAFT_gp12')\n",
    "Clade(branch_length=0.0, name='Wile-DRAFT_gp11')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Clade(branch_length=0.00623, name='Vix-DRAFT_gp11')\n",
    "Clade(branch_length=0.0, name='Microwolf-DRAFT_gp14')\n",
    "Clade(branch_length=0.0, name='205870_14')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(sea_inphinity.get_list_genes(['Vix-DRAFT_gp11'], True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sea_inphinity.print_informations_on_phage('Vix-DRAFT_gp11')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sea_inphinity.print_informations_on_phages(['Vix-DRAFT_gp11', 'Microwolf-DRAFT_gp14', '205870_14'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sea_inphinity.print_informations_on_phages(['663557_12', 'Backyardigan-DRAFT_gp12', 'Wile-DRAFT_gp11'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sea_inphinity.print_informations_on_phages(sea_inphinity.leaves)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Checking Theory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creation of the FASTA file\n",
      "Alignment with MUSCLE\n",
      "Compute tree\n",
      "Prepare Tree Fig\n",
      "===========================\n",
      "\u001b[34mBlue: Host come from the original database\u001b[0m\n",
      "\u001b[31mRed: Host come from GenBank\u001b[0m\n",
      "\u001b[33mYellow: DRAFT\u001b[0m\n",
      "===========================\n",
      "Mycobacterium phage Pukovnik\n",
      "********************************             \n",
      "Gene ID: 540068_15 \n",
      "Phage ID: 540068 \n",
      "Phage Name: Pukovnik \n",
      "Phage Accession: EU744250.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Pukovnik\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Turbido-DRAFT_gp14 \n",
      "Phage ID: Turbido-DRAFT \n",
      "Phage Name: Turbido-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage D29\n",
      "********************************             \n",
      "Gene ID: 28369_14 \n",
      "Phage ID: 28369 \n",
      "Phage Name: D29 \n",
      "Phage Accession: AF022214.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage D29\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Che12\n",
      "********************************             \n",
      "Gene ID: 373405_14 \n",
      "Phage ID: 373405 \n",
      "Phage Name: Che12 \n",
      "Phage Accession: DQ398043.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Che12\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: George-DRAFT_gp9 \n",
      "Phage ID: George-DRAFT \n",
      "Phage Name: George-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Bxb1\n",
      "********************************             \n",
      "Gene ID: 148603_11 \n",
      "Phage ID: 148603 \n",
      "Phage Name: Bxb1 \n",
      "Phage Accession: AF271693.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Bxb1\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Bxz2\n",
      "********************************             \n",
      "Gene ID: 205870_14 \n",
      "Phage ID: 205870 \n",
      "Phage Name: Bxz2 \n",
      "Phage Accession: AY129332.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Bxz2\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: JHC117-DRAFT_gp14 \n",
      "Phage ID: JHC117-DRAFT \n",
      "Phage Name: JHC117-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Microwolf-DRAFT_gp14 \n",
      "Phage ID: Microwolf-DRAFT \n",
      "Phage Name: Microwolf-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Vix-DRAFT_gp11 \n",
      "Phage ID: Vix-DRAFT \n",
      "Phage Name: Vix-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "********************************             \n",
      "Gene ID: 663557_12 \n",
      "Phage ID: 663557 \n",
      "Phage Name: Peaches \n",
      "Phage Accession: GQ303263.1\n",
      "\u001b[34mHostStrin: Mycobacterium smegmatis mc2 155\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Backyardigan-DRAFT_gp12 \n",
      "Phage ID: Backyardigan-DRAFT \n",
      "Phage Name: Backyardigan-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Wile-DRAFT_gp11 \n",
      "Phage ID: Wile-DRAFT \n",
      "Phage Name: Wile-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage DD5\n",
      "********************************             \n",
      "Gene ID: 540064_13 \n",
      "Phage ID: 540064 \n",
      "Phage Name: DD5 \n",
      "Phage Accession: EU744252.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage DD5\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Lockley\n",
      "********************************             \n",
      "Gene ID: 540067_12 \n",
      "Phage ID: 540067 \n",
      "Phage Name: Lockley \n",
      "Phage Accession: EU744249.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Lockley\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Jasper\n",
      "********************************             \n",
      "Gene ID: 540065_13 \n",
      "Phage ID: 540065 \n",
      "Phage Name: Jasper \n",
      "Phage Accession: EU744251.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Jasper\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage U2\n",
      "********************************             \n",
      "Gene ID: 260120_gp11 \n",
      "Phage ID: 260120 \n",
      "Phage Name: U2 \n",
      "Phage Accession: AY500152.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage U2\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Doom-DRAFT_gp14 \n",
      "Phage ID: Doom-DRAFT \n",
      "Phage Name: Doom-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: SargentShorty9-DRAFT_gp12 \n",
      "Phage ID: SargentShorty9-DRAFT \n",
      "Phage Name: SargentShorty9-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Bethlehem\n",
      "********************************             \n",
      "Gene ID: 260121_gp11 \n",
      "Phage ID: 260121 \n",
      "Phage Name: Bethlehem \n",
      "Phage Accession: AY500153.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Bethlehem\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage KBG\n",
      "********************************             \n",
      "Gene ID: 540066_KBG_13 \n",
      "Phage ID: 540066 \n",
      "Phage Name: KBG \n",
      "Phage Accession: EU744248.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage KBG\u001b[0m\n",
      "********************************\n",
      "********************************             \n",
      "Gene ID: 555603_12 \n",
      "Phage ID: 555603 \n",
      "Phage Name: Solon \n",
      "Phage Accession: EU826470.1\n",
      "\u001b[34mHostStrin: Mycobacterium smegmatis mc2 155\u001b[0m\n",
      "********************************\n",
      "********************************             \n",
      "Gene ID: 701456_14 \n",
      "Phage ID: 701456 \n",
      "Phage Name: SkiPole \n",
      "Phage Accession: GU247132.1\n",
      "\u001b[34mHostStrin: Mycobacterium smegmatis mc2 155\u001b[0m\n",
      "********************************\n",
      "********************************             \n",
      "Gene ID: 701456_87 \n",
      "Phage ID: 701456 \n",
      "Phage Name: SkiPole \n",
      "Phage Accession: GU247132.1\n",
      "\u001b[34mHostStrin: Mycobacterium smegmatis mc2 155\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage Omega\n",
      "********************************             \n",
      "Gene ID: 205879_166 \n",
      "Phage ID: 205879 \n",
      "Phage Name: Omega \n",
      "Phage Accession: AY129338.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage Omega\u001b[0m\n",
      "********************************\n",
      "Mycobacterium phage L5\n",
      "********************************             \n",
      "Gene ID: 31757_gene14 \n",
      "Phage ID: 31757 \n",
      "Phage Name: L5 \n",
      "Phage Accession: Z18946.1\n",
      "\u001b[31mHostStrin: Mycobacterium phage L5\u001b[0m\n",
      "********************************\n",
      "********************************             \n",
      "Gene ID: 711470_16 \n",
      "Phage ID: 711470 \n",
      "Phage Name: RedRock \n",
      "Phage Accession: GU339467.1\n",
      "\u001b[34mHostStrin: Mycobacterium smegamatis mc2 155\u001b[0m\n",
      "********************************\n",
      "Not Found\n",
      "********************************             \n",
      "Gene ID: Trixie-DRAFT_gp17 \n",
      "Phage ID: Trixie-DRAFT \n",
      "Phage Name: Trixie-DRAFT \n",
      "Phage Accession: \n",
      "\u001b[33mHostStrin: Not Found\u001b[0m\n",
      "********************************\n"
     ]
    }
   ],
   "source": [
    "sea_inphinity.verbose = False\n",
    "sea_inphinity.build_tree('2799')\n",
    "sea_inphinity.print_informations_on_phages(sea_inphinity.leaves)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "phage = 'George-DRAFT'\n",
    "print(phage.split('-')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
